go/api/config.go

```diff
@@ -0,0 +1,309 @@
+	if addr.Host == "" {
+		addr.Host = defaultHost
+	}
+
+	if addr.Port == 0 {
+		addr.Port = defaultPort
+	}
+
+	return &protobuf.NodeAddress{Host: addr.Host, Port: uint32(addr.Port)}
+}
+
+// RedisCredentials represents the credentials for connecting to a Redis server.
+type RedisCredentials struct {
+	// Optional: the username that will be used for authenticating connections to the Redis servers. If not supplied, "default"
+	// will be used.
+	username string
+	// Required: the password that will be used for authenticating connections to the Redis servers.
+	password string
+}
+
+// NewRedisCredentials returns a new RedisCredentials struct with the given username and password.
+func NewRedisCredentials(username string, password string) *RedisCredentials {
+	return &RedisCredentials{username, password}
+}
+
+// NewRedisCredentialsWithDefaultUsername returns a new RedisCredentials struct with a default username of "default" and the
+// given password.
+func NewRedisCredentialsWithDefaultUsername(password string) *RedisCredentials {
+	return &RedisCredentials{password: password}
+}
+
+func (creds *RedisCredentials) toProtobuf() *protobuf.AuthenticationInfo {
+	return &protobuf.AuthenticationInfo{Username: creds.username, Password: creds.password}
+}
+
+// ReadFrom represents the client's read from strategy.
+type ReadFrom int
+
+const (
+	// Primary - Always get from primary, in order to get the freshest data.
+	Primary ReadFrom = iota
+	// PreferReplica - Spread the requests between all replicas in a round-robin manner. If no replica is available, route the
+	// requests to the primary.
+	PreferReplica
+)
+
+func mapReadFrom(readFrom ReadFrom) protobuf.ReadFrom {
+	if readFrom == PreferReplica {
+		return protobuf.ReadFrom_PreferReplica
+	}
+
+	return protobuf.ReadFrom_Primary
+}
+
+type baseClientConfiguration struct {
+	addresses      []NodeAddress
+	useTLS         bool
+	credentials    *RedisCredentials
+	readFrom       ReadFrom
+	requestTimeout int
+	clientName     string
+}
+
+func (config *baseClientConfiguration) toProtobuf() *protobuf.ConnectionRequest {
+	request := protobuf.ConnectionRequest{}
+	for _, address := range config.addresses {
+		request.Addresses = append(request.Addresses, address.toProtobuf())
+	}
+
+	if config.useTLS {
+		request.TlsMode = protobuf.TlsMode_SecureTls
+	} else {
+		request.TlsMode = protobuf.TlsMode_NoTls
+	}
+
+	if config.credentials != nil {
+		request.AuthenticationInfo = config.credentials.toProtobuf()
+	}
+
+	request.ReadFrom = mapReadFrom(config.readFrom)
+	if config.requestTimeout != 0 {
+		request.RequestTimeout = uint32(config.requestTimeout)
+	}
+
+	if config.clientName != "" {
+		request.ClientName = config.clientName
+	}
+
+	return &request
+}
+
+// BackoffStrategy represents the strategy used to determine how and when to reconnect, in case of connection failures. The
+// time between attempts grows exponentially, to the formula:
+//
+//	rand(0 ... factor * (exponentBase ^ N))
```

Comment:
In [Go docs](https://tip.golang.org/doc/comment), a code block is a span of indented lines, which I've used here for the formula. Go docs support paragraphs, headings, links, lists, and preformatted code blocks. They do not seem to have bold, italics, or inline code highlighting.